src/Microsoft.ML.Core/Utilities/ThreadUtils.cs

```diff
@@ -4,44 +4,154 @@
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task Queue(Action<object> threadStart, object state) => Queue((Delegate)threadStart, state);
+
+            private static Task Queue(Delegate threadStart, object state)
             {
-                IsBackground = false
-            };
+                // Create the TaskCompletionSource used to represent this work.
+                // Call sites only care about completion, not about the distinction between
+                // success and failure and do not expect exceptions to be propagated in this manner,
+                // so only SetResult is used.
+                var tcs = new TaskCompletionSource<bool>(TaskContinuationOptions.RunContinuationsAsynchronously);
+
+                // Queue the work for a thread to pick up.  If no thread is immediately available, create one.
+                if (!Enqueue((threadStart, state, tcs)))
+                {
+                    CreateThread();
+                }
+
+                // Return the task.
+                return tcs.Task;
+
+                void CreateThread()
+                {
+                    // Create a new background thread to run the work.
+                    var t = new Thread(() =>
+                    {
+                        // Repeatedly get the next item and invoke it, setting its TCS when we're done.
+                        // This will wait for up to the idle time before giving up and exiting.
+                        while (TryDequeue(out (Delegate, object, TaskCompletionSource<bool>) item))
+                        {
+                            try
+                            {
+                                if (item.Item1 is Action<object> pts)
+                                {
+                                    pts(item.Item2);
+                                }
+                                else
+                                {
+                                    ((Action)item.Item1)();
+                                }
+                            }
+                            finally
+                            {
+                                item.Item3.SetResult(true);
+                            }
+                        }
+                    });
+                    t.IsBackground = true;
+                    t.Start();
+                }
+
+                bool Enqueue((Delegate, object, TaskCompletionSource<bool>) item)
+                {
+                    // Enqueue the work.  If there are currently fewer threads waiting
+                    // for work than there are work items in the queue, create another
+                    // thread.  This is a heuristic, in that we might end up creating
```

Comment:
"create another                     // thread." -> "return false to indicate a new thread should be created"
but perhaps it would be clearer to create the thread within Enqueue?

Reply:
Ok